bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/BookKeeperBuilder.java

```diff
@@ -0,0 +1,104 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.client.api;
+
+import java.io.IOException;
+import io.netty.channel.EventLoopGroup;
+import io.netty.util.HashedWheelTimer;
+import org.apache.bookkeeper.feature.FeatureProvider;
+import org.apache.bookkeeper.net.DNSToSwitchMapping;
+import org.apache.bookkeeper.stats.StatsLogger;
+import org.apache.zookeeper.ZooKeeper;
+
+/**
+ * BookKeeper Client Builder to build client instances.
+ *
+ * @since 4.6
+ */
+public interface BookKeeperBuilder {
+
+    /**
+     * Configure the bookkeeper client with a provided Netty EventLoopGroup.
+     *
+     * @param eventLoopGroup an external {@link EventLoopGroup} to use by the bookkeeper client.
+     *
+     * @return client builder.
+     */
+    BookKeeperBuilder eventLoopGroup(EventLoopGroup eventLoopGroup);
+
+    /**
+     * Configure the bookkeeper client with a provided ZooKeeper client.
+     *
+     * @param zk an external {@link ZooKeeper} client to use by the bookkeeper client.
+     *
+     * @return client builder.
+     */
+    BookKeeperBuilder zk(ZooKeeper zk);
```

Comment:
@sijie  I am not sure we really want this in the new API.
It would be more useful ZKClientConfig (an in the near future I will request and push a change to have it). Because our (BK) ZooKeeper client is better, as it is a wrapper for bare ZooKeeper client

Reply:
@eolivelli I think it is fine to have this in a builder.  the `#zk(...)` is to pass an established zookeeper client, the configuration already has zookeeper connection parameters. you don't need additional `ZKClientConfig`. the `#zk(...)` is useful for sharing zookeeper instance between bookkeeper client and applications (such as message brokers).